File: scripts/excel2xml.py
#!/usr/bin/env python
"""
* Requires xlrd.
$ python xl.py [filename].xsl
"""
import os
import sys
import xlrd
import time

from xlrd import open_workbook

if len(sys.argv) < 2:
	print "Error: You must supply a filename"
	sys.exit(1)

filename = sys.argv[-1]
if os.path.isfile(filename):
#	pathname=os.path.split(filename)[0]
	filebasename=os.path.split(filename)[1]
	xmlfilename=os.path.splitext(filebasename)[0]+".xml"
	with open(xmlfilename,"w") as xmlfile:
		print "####################################################################"
		print "Converting from:"+filename
		print "Converting to:  "+xmlfilename
		print "####################################################################"
		wb=open_workbook(filename)
		###########################################
		### Get the sheet names and loop over them
		###########################################
		names=wb.sheet_names()
		xmlfile.write("<!-- WorkBook Path: "+filename+" -->\n")
		xmlfile.write("<!-- Date: "+time.strftime("%c")+" -->\n")
		xmlfile.write("<Workbook>\n")
		###########################################
		### What sheet is the testplan on?
		###########################################
		for (i,name) in enumerate(names):
			s=wb.sheet_by_index(i)
			###########################################
			### Get the row that has the first data
			###########################################
			first_row=0
			row=0
			while row < s.nrows and first_row==0:
				values=s.row_values(row)
				for (i, item) in enumerate(values):
					if item<>'' and first_row==0:
						first_row=row
				row=row+1
			###########################################
			### Get the col that has the first data
			###########################################
			first_col=0
			col=0
			while col < s.ncols and first_col==0:
				values=s.col_values(col)
				for (i, item) in enumerate(values):
					if item<>'' and first_col==0:
						first_col=col
				col = col + 1
			###########################################
			### print out XML for the testplan, ignore
			### the other sheets
			###########################################
			if s.nrows-first_row-1>0 and s.ncols-1-first_col>0:
				print "Found Worksheet:",name
				xmlfile.write("  <Worksheet>  <!-- "+name+" -->\n")
				xmlfile.write("    <Table>\n")
				################################################
				### The first row has the headings, ignore them
				################################################
				for row in range(first_row+1,s.nrows):
					xmlfile.write( "      <Row>\n")
					for col in range(first_col,s.ncols):
						line="        <Cell>"+str(s.cell(row,col).value)+"</Cell> <!-- "+str(s.cell(first_row,col).value)+" -->\n"
						xmlfile.write(line)
					xmlfile.write( "      </Row>\n")
				xmlfile.write( "    </Table>\n")
				xmlfile.write( "  </Worksheet>\n")
		xmlfile.write( "</Workbook>\n")
		print "Done!"
else:
	sys.stderr.write("Error: "+filename+" doesn't exist!\n" )
	sys.stderr.write("Usage: excel2xml.py <excelfilename.xls/xlsx>\n" )
	sys.exit(2)

File: scripts/excel2xml.pl
#!/usr/bin/perl
use strict;
use Getopt::Long;
use Spreadsheet::ParseExcel;
use Spreadsheet::XLSX;
use POSIX qw(strftime);
use File::Basename;

my $worksheet;

my $excelfile;
my $datestring = strftime "%Y %b %e %H:%M:%S", localtime;

GetOptions ("input|i=s"			=> \$excelfile	)	or die;

defined $excelfile or die "You must provide an excelfile name with -i or -input!\n";

(my $filename,my $path,my $suffix) = fileparse($excelfile,qr"\..[^.]*$");
my $xmlfilename="$filename.xml";

print "##################################################\n";
print "Converting from:$excelfile\n";
print "Converting to:  $xmlfilename\n";
print "##################################################\n";

open XMLFILE, ">$xmlfilename" or die $!;
my $parser;
my $workbook;

if ($suffix =~ /xlsx/) {
	$workbook   = Spreadsheet::XLSX->new($excelfile);
} elsif ($suffix =~ /xls/) {
	$parser   = Spreadsheet::ParseExcel->new();
	$workbook = $parser->parse($excelfile);
	die $parser->error(), ".\n" if ( !defined $workbook );
}


print XMLFILE "<!-- WorkBook Path: $excelfile -->\n";
print XMLFILE "<!-- Date: $datestring -->\n";
print XMLFILE "<Workbook>\n";
for my $worksheet ( $workbook->worksheets() ) {
	my $worksheetname=$worksheet->get_name();
 	my ( $row_min, $row_max ) = $worksheet->row_range();
   my ( $col_min, $col_max ) = $worksheet->col_range();
	##################################################################
	## Not necessary that the first row is the one as row_min since ##
	## the user might have backannotated coverage numbers           ##
	## So look for the first row that has more than 4 columns       ##
	##################################################################
	my $start_row=$row_min;
	for (my $r=$row_min; $r<$row_max;$r++) {
		my $cell_with_contents=0;
		for (my $c=$col_min;$c<$col_max;$c++) {
   		my $cell = $worksheet->get_cell($r, $c);
			next unless $cell;
			my $val=$cell->value();
			if ($val ne "") {
			  	$cell_with_contents++;
			}
		}
		if ($cell_with_contents>4) {
			$start_row=$r;
			last;
		}
	}

	# Skip the worksheet with _LINK_DETAILS (hidden one) and the ones with columns less than 6 (should make it only the verification plan)
	if ( !($worksheetname =~ /^_/) && $col_max-$col_min>5) {
	   print "Found worksheet: $worksheetname\n";
		print "The rows range from $start_row to $row_max  \n";
		print "The columns range from $col_min to $col_max\n";
		print "These are the columns:\n";
		for (my $c=$col_min;$c<=$col_max;$c++) {
			my $description= $worksheet->get_cell($start_row, $c)->value();
			printf "Column:%2d  -- $description\n",$c;
		}

	   print "##################################################\n";
		print XMLFILE "   <Worksheet>  <!-- $worksheetname -->\n";
		print XMLFILE "     <Table>\n";
		for (my $row=$start_row+1;$row<=$row_max;$row++) {
			print XMLFILE "        <Row>\n";
			for (my $col=$col_min;$col<=$col_max;$col++) {
				print XMLFILE "           <Cell>";
				my $cell = $worksheet->get_cell($row, $col);
				if ($cell) {
					my $value = $cell->value();
					# Since we are writing out XML we need to substitute some reserved characters
					$value =~ s/&/&amp;/g;
					$value =~ s/\"/&quot;/g;
					$value =~ s/</&lt;/g;
					$value =~ s/>/&gt;/g;
					print XMLFILE "$value";
				}
				my $description= $worksheet->get_cell($start_row, $col)->value();
				# Since we are writing out XML we need to substitute some reserved characters
				$description =~ s/&/&amp;/g;
				$description =~ s/\"/&quot;/g;
				$description =~ s/</&lt;/g;
				$description =~ s/>/&gt;/g;
				print XMLFILE "</Cell>  <!-- $description -->\n";
			}
			print XMLFILE "        </Row>\n";
		}
		print XMLFILE "     </Table>\n";
		print XMLFILE "   </Worksheet>  <!-- $worksheetname -->\n";
	}
}
print XMLFILE "</Workbook>\n";
close (XMLFILE);

print "Converted $excelfile to $xmlfilename\n";

